### PR TITLE
LRDOCS-9788 use general token variables to set DXP/Portal values on …

### PR DESCRIPTION
**Pull request by jhinkey**
…the site's articles and projects

I propose applying general values to projects that use Portal values, since features that work with Portal also work with DXP. Projects that are DXP only ALWAYS use DXP values.

To apply DXP generally, use these settings in `_common.sh`:

```
readonly LIFERAY_LEARN_GENERAL_API_TOKEN="${LIFERAY_LEARN_DXP_API_TOKEN}"
readonly LIFERAY_LEARN_GENERAL_DOCKER_IMAGE_VALUE="${LIFERAY_LEARN_DXP_DOCKER_IMAGE_VALUE}"
readonly LIFERAY_LEARN_GENERAL_WORKSPACE_TOKEN="${LIFERAY_LEARN_DXP_WORKSPACE_TOKEN}"
```

To apply Portal generally, use these settings in `_common.sh`:

```
readonly LIFERAY_LEARN_GENERAL_API_TOKEN="${LIFERAY_LEARN_PORTAL_API_TOKEN}"
readonly LIFERAY_LEARN_GENERAL_DOCKER_IMAGE_VALUE="${LIFERAY_LEARN_PORTAL_DOCKER_IMAGE_VALUE}"
readonly LIFERAY_LEARN_GENERAL_WORKSPACE_TOKEN="${LIFERAY_LEARN_PORTAL_WORKSPACE_TOKEN}"
```

Run build_site.sh on a .md file.

- The `build.gradle` and `gradle.properties` files output in the project .zip file in `build/output/` have DXP values if you set the GENERAL vars to use DXP or have Portal values if you set the GENERAL vars to use Portal.
- The article html files in `build/output/` use the respective DXP/Portal docker image

### CODE REVIEW

File: _common.sh [hunk 1]
```diff
@@ -5,16 +5,23 @@ readonly LIFERAY_LEARN_COMMERCE_DOCKER_IMAGE_VALUE=liferay\\/commerce\\:2.0.5
...
 readonly LIFERAY_LEARN_PORTAL_DOCKER_IMAGE_TOKEN=\\[\$LIFERAY_LEARN_PORTAL_DOCKER_IMAGE\$\\]
 readonly LIFERAY_LEARN_PORTAL_DOCKER_IMAGE_VALUE=liferay\\/portal\\:7.4.1-ga2
 readonly LIFERAY_LEARN_PORTAL_GIT_TAG_TOKEN=\\[\$LIFERAY_LEARN_PORTAL_GIT_TAG\$\\]
 readonly LIFERAY_LEARN_PORTAL_GIT_TAG_VALUE=7.4.1-ga2
 readonly LIFERAY_LEARN_PORTAL_WORKSPACE_TOKEN=portal-7.4-ga2
 
+readonly LIFERAY_LEARN_GENERAL_API_TOKEN="${LIFERAY_LEARN_PORTAL_API_TOKEN}"
+readonly LIFERAY_LEARN_GENERAL_DOCKER_IMAGE_VALUE="${LIFERAY_LEARN_PORTAL_DOCKER_IMAGE_VALUE}"
+readonly LIFERAY_LEARN_GENERAL_DOCKER_IMAGE_TOKEN=\\[\$LIFERAY_LEARN_GENERAL_DOCKER_IMAGE\$\\]
+readonly LIFERAY_LEARN_GENERAL_WORKSPACE_TOKEN="${LIFERAY_LEARN_PORTAL_WORKSPACE_TOKEN}"
```
Comment (jhinkey) [inline _common.sh:23]: Having the Workspace values in this file is helpful. ... The version format is slightly different from the Git tag and Docker tag.

**Comment by jhinkey** [comment]
@rbohl @jrhoun @sez11a - I tried combining the PORTAL and GENERAL variables, but there is an issue ...

The [Starting with a Docker Image#Get Started with Liferay Portal section](https://github.com/liferay/liferay-learn/blob/master/docs/dxp/latest/en/getting-started/starting-with-a-docker-image.md#get-started-with-liferay-portal) references a variable that must point to the most recent **portal Docker image**. So we must keep the PORTAL_DOCKER_IMAGE token and value variables. And we still need GENERAL_DOCKER_IMAGE variables to apply DXP/Portal images generally.

Importantly, having Portal Docker image tag, Workspace product, and API variables creates a consistent intuitive way to update DXP and Portal values and to switch between DXP and Portal. 
   1. Update the DXP and PORTAL values for the respective variables.
   1. Assign the GENERAL variables to the DXP variables or to the PORTAL variables.

So I'm sticking with this proposal. ;-)

**Comment by jhinkey** [comment]
@sez11a @rbohl @jrhoun

Also, I think these variables would be much easier to distinguish if we removed the `LIFERAY_LEARN_` prefix from them. Any reasons to keep the prefix?

**Comment by rbohl** [comment]
If I recall correctly, BChan wanted the LIFERAY_LEARN prefix

**Comment by rbohl** [comment]
Merged locally.

**Comment by rbohl** [comment]
@jhinkey what's going to cause us to need to switch between DXP and PORTAL in the future? I know we just had to switch back form using the DXP tokens, but in the future will we actually need to do this?

I'm thinking through the 7.4 release cycle, using one of the articles that currently relies on your GENERAL token: [Form Storage Adapters](https://learn.liferay.com/dxp/latest/en/process-automation/forms/developer-guide/writing-a-form-storage-adapter.html).

Why would we ever again need to switch this to a DXP value at any time during the 7.4 release cycle?

As soon as we do that no CE user can ever follow the tutorial instructions, right?

**Comment by jhinkey** [comment]
@rbohl I'm thinking more along the lines of when we go to 7.5+. When the first 7.5 Portal GA is released, we'll want to ...

1. Update the PORTAL values in _common.sh to use the 7.5 values.
2. Switch our GENERAL tokens from using the DXP value to using the PORTAL values.

**Comment by jhinkey** [comment]
Our general tutorials demonstrate using the _latest_, cutting edge release. The understanding is that the audience can adapt the release values to whatever 7.2+ edition or version they want. Here's an example sequence of changing the GENERAL values...

1. Right now, use PORTAL (7.4) values since the latest portal 7.4 release is the cutting edge DXP/Portal release.
2. When DXP 7.4 releases, set GENERAL to the latest DXP 7.4 release
3. When Portal 7.5 releases, set GENERAL to that Portal release
4. When DXP 7.5 releases, set GENERAL to that DXP release
...

**Comment by rbohl** [comment]
If we use steps 2 and 5 the tutorial instructions will still work for community members (without a DXP license)?

**Comment by jhinkey** [comment]
The assumption is that DXP releases are happening so frequently that the licenses in the Docker images still have some free trial time.

IF a trial license expires for the latest DXP image, they'll have to either pass a new trial license to the image/container OR they bail on DXP and use the comparable Portal image.

**Comment by jhinkey** [comment]
... But once again, the guiding principle is to always demonstrate the most cutting edge release.

**Comment by sez11a** [comment]
@jhinkey and @rbohl, I think there may be a misunderstanding here. We NEVER want to switch wholesale to DXP anymore. We want to use DXP only for those features that REQUIRE DXP; everything else should use Portal. Therefore, I don't think we need general variables at all. Each article should specify the Portal or the DXP variable that's the most current.

**Pull request closed without merging**